lightning-invoice/src/payment.rs

```diff
@@ -805,6 +848,50 @@ mod tests {
 		assert_eq!(*payer.attempts.borrow(), 3);
 	}
 
+	#[test]
+	fn fails_paying_invoice_after_max_retry_timeout() {
+		let event_handled = core::cell::RefCell::new(false);
+		let event_handler = |_: &_| { *event_handled.borrow_mut() = true; };
+
+		let payment_preimage = PaymentPreimage([1; 32]);
+		let invoice = invoice(payment_preimage);
+		let final_value_msat = invoice.amount_milli_satoshis().unwrap();
+
+		let payer = TestPayer::new()
+			.expect_send(Amount::ForInvoice(final_value_msat))
+			.expect_send(Amount::OnRetry(final_value_msat / 2));
+
+		let router = TestRouter {};
+		let scorer = RefCell::new(TestScorer::new());
+		let logger = TestLogger::new();
+		let invoice_payer =
+			InvoicePayer::new(&payer, router, &scorer, &logger, event_handler, RetryAttempts::MaxDuration(Duration::from_millis(100)));
+
+		let payment_id = Some(invoice_payer.pay_invoice(&invoice).unwrap());
+		assert_eq!(*payer.attempts.borrow(), 1);
+
+		let event = Event::PaymentPathFailed {
+			payment_id,
+			payment_hash: PaymentHash(invoice.payment_hash().clone().into_inner()),
+			network_update: None,
+			rejected_by_dest: false,
+			all_paths_failed: true,
+			path: TestRouter::path_for_value(final_value_msat),
+			short_channel_id: None,
+			retry: Some(TestRouter::retry_for_invoice(&invoice)),
+		};
+		invoice_payer.handle_event(&event);
+		assert_eq!(*event_handled.borrow(), false);
+		assert_eq!(*payer.attempts.borrow(), 2);
+
+    std::thread::sleep(Duration::from_millis(101));
```

Comment:
Sleeps in testing tend to be a bit flaky and slow down test execution. Ideally, we can inject the time into the code such that in testing it can be manually advanced. We do that in the scorer code:

https://github.com/lightningdevkit/rust-lightning/blob/a3832b029c151e2afd910f5372d1aadcd4e394bc/lightning/src/routing/scoring.rs#L1164

We wouldn't need to expose the `Time` trait in `InvoicePayer`'s public interface since it is an implementation detail and can be conditioned for tests to use `SinceEpoch` instead of `SystemTime`. It would be great if we can refactor the time code into a `util::time` module such that it could be reused here. We could symlink it into the `lightning-invoice` crate to re-use across crates.